Extendtci command as separate branch

hpya93 opened the pull request:
Hi @jhand2 
Have added ExtendTCI command tests in separate branch
- cross checking current and cumulative measurement - using GettaggedTci and CertifyKey returned DPE leaf cert's MultiTcbInfo extension
- testing by deriving a child context and extending values to derived context - and cleanup derived context.

jhand2 commented on verification/abi.go:
```diff
@@ -3,11 +3,12 @@
 	"reflect"
 )
 
+var DefaultContextHandle = ContextHandle{0}
```
nit: I think this can be const

hpya93 replied:
Checked this, seems golang doesnt allow arrays to be constants.

jhand2 commented on verification/tagTCI.go:
```diff
@@ -3,65 +3,227 @@
+}
+
+// Checks whether error is reported when the caller from one locality makes call to another locality.
+func testTagTCIWrongLocality(d TestDPEInstance, c DPEClient, t *testing.T, handle *ContextHandle) {
```
Maybe we should just make each of these a separate test since they're testing difference cases.

hpya93 replied:
I shall add a BasicTestcase that checks common things like Wrong locality and Invalid handle errors

hpya93 replied:
Added the func and raised PR. link https://github.com/chipsalliance/caliptra-dpe/pull/255

jhand2 commented on verification/client.go:
```diff
@@ -33,14 +33,24 @@ type DPETCI struct {
+	SignatureS       []byte
+}
+
 type DPEClient interface {
 	InitializeContext(flags InitCtxFlags) (*ContextHandle, error)
 	GetProfile() (*GetProfileResp, error)
+	DeriveChild(handle *ContextHandle, inputData []byte, flags DeriveChildFlags, tciType uint32, targetLocality uint32) (*DeriveChildResp, error)
 	CertifyKey(handle *ContextHandle, label []byte, format CertifyKeyFormat, flags CertifyKeyFlags) (*CertifiedKey, error)
 	GetCertificateChain() ([]byte, error)
 	TagTCI(handle *ContextHandle, tag TCITag) (*ContextHandle, error)
 	GetTaggedTCI(tag TCITag) (*DPETCI, error)
+	ExtendTCI(handle *ContextHandle, inputData []byte) (*ContextHandle, error)
 	DestroyContext(handle *ContextHandle, flags DestroyCtxFlags) error
+	RotateContextHandle(handle *ContextHandle, flags RotateContextHandleFlags) (*ContextHandle, error)
+	Sign(handle *ContextHandle, label []byte, flags SignFlags, toBeSigned []byte) (*DPESignedHash, error)
 }
```
Could you add all the client code in one separate PR without the tests? All of the other PRs you've sent depend on it, so I think it will make all of the PRs quite a bit smaller.

hpya93 replied:
sure I shall raise one

hpya93 replied:
Please find PR link here: https://github.com/chipsalliance/caliptra-dpe/pull/255

hpya93 commented:
Have rebased this branch with respect to PR 255 for easier comparison.

jhand2 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,209 @@
+		t.Errorf("Child node's cumulative TCI %x, expected %x", childTci.CumulativeTCI, wantCumulativeTCI)
+	}
+
+	// Extend TCI to child context and check TCI_CURRENT and TCI_CUMULATIVE
+	_, err = c.ExtendTCI(newHandle, extendTciValue)
+	if err != nil {
+		t.Fatalf("Could not tag TCI: %v", err)
+	}
+
+	childExtendTci, err := c.GetTaggedTCI(childTag)
+	if err != nil {
+		t.Fatalf("Could not get tagged TCI: %v", err)
+	}
+
+	if !bytes.Equal(childExtendTci.CurrentTCI, extendTciValue) {
+		t.Errorf("GetTaggedTCI returned current TCI %x, expected %x", childExtendTci.CurrentTCI, extendTciValue)
+	}
+
+	// Check TCI_CUMULATIVE after extending input to child context
+	hasher.Reset()
+	hasher.Write(childTci.CumulativeTCI)
+	hasher.Write(childExtendTci.CurrentTCI)
+	wantCumulativeTCI = hasher.Sum(nil)
+	if !bytes.Equal(childExtendTci.CumulativeTCI, wantCumulativeTCI) {
+		t.Errorf("Child node's cumulative TCI %x, expected %x", childExtendTci.CumulativeTCI, wantCumulativeTCI)
+	}
```
This could be a 3rd test case (since it requires `ExtendTCI` support, but the earlier part of the test doesn't)

hpya93 replied:
Have split into 3rd test case

hpya93 commented on verification/verification.go:
```diff
@@ -56,9 +56,9 @@ var UnsupportedCommandFlag = TestCase{
 
 var AllTestCases = []TestCase{
 	CertifyKeyTestCase,
+	TagTCITestCase,
```
The CertifyKey_simulation interfered with TAGTCI when testing with Derived context. Added DestroyContext in  CertifyKey_simulation testcase to handle this.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,209 @@
+
+// Checks whether the default handle remains unchanged when extended with TCI value.
+func testTagTCIHandle(d TestDPEInstance, c DPEClient, t *testing.T, handle *ContextHandle, digestLen int, tag TCITag) {
+	tciValue := make([]byte, digestLen)
+	for i := range tciValue {
+		tciValue[i] = byte(i)
+	}
```
Yes, it was a mistake.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,209 @@
+}
+
+// Checks whether the default handle remains unchanged when extended with TCI value.
+func testTagTCIHandle(d TestDPEInstance, c DPEClient, t *testing.T, handle *ContextHandle, digestLen int, tag TCITag) {
```
I thought to have modular functions to improve readability. Just tried to have a track of the tests by just going through the function calls. If you insist I shall move to caller.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,209 @@
+
+	// Make sure the current TCI is updated
+	// For profiles which use auto-initialization, we don't know the expected TCIs so we can check only current TCI
+	testCurrentAndCumulativeTCI(d, c, t, handle, digestLen, tag)
+
+	// Make sure some other tag is still not found.
+	if _, err := c.GetTaggedTCI(TCITag(98765)); !errors.Is(err, StatusBadTag) {
+		t.Fatalf("GetTaggedTCI returned %v, want %v", err, StatusBadTag)
+	}
+
+	// Pass TCI input to DerivedChild context and check cumulative TCI
+	testForDerivedChildContexts(d, c, t, handle, digestLen, tag)
```
Ok, In that case I shall write a separate file for extendTCI command instead of clubbing it with TagTci.go

hpya93 commented on verification/certifyKey.go:
```diff
@@ -433,9 +433,11 @@ func checkCertificateStructure(t *testing.T, certBytes []byte) *x509.Certificate
+		if simulation {
+			client.DestroyContext(ctx, 0)
+		}
+	}()
```
Fixed code here to place TagTCI testcase after CertifyKey_Simulation testcase in `AllTestCases` list. 

This is required because handle created in CertifyKey_Simulation mode interfered with derived context created for TagTCI tests with below error. 
```
=== RUN   TestRunAll/DefaultSupport-ExtendDerivedTciTestCase
    tagTCI.go:171: [FATAL]: Error while creating default child handle in default context: argument is invalid
```

hpya93 commented on verification/certifyKey.go:
```diff
@@ -449,7 +451,7 @@ func testCertifyKey(d TestDPEInstance, client DPEClient, t *testing.T, simulatio
 
 	seqLabel := make([]byte, digestLen)
-	for i, _ := range seqLabel {
+	for i := range seqLabel {
```
IDE was giving a warning for empty value "simplify range expression", so removed '_'

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,218 @@
-	ctx := getInitialContextHandle(d, client, t, useSimulation)
+const defaultCtxTCITag = TCITag(12345)
+const nonExistentTCITag = TCITag(98765)
+const childCtxTCITag = TCITag(34567)
```
Added constants for better readability

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,218 @@
+		t.Errorf("New context handle from TagTCI was %x, expected %x", newHandle, handle)
+	}
+
+	// Retag a tagged TCI should report error
```
Added this new check, it checks if retagging a tagged TCI returns proper error code.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,218 @@
+
+// Check tagTCI command with default context handle.
+func TestTagTCI(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
1. Removed clubbed test functions to put them as separate test cases, updated comments

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,218 @@
+
+// Check whether the ExtendTCI command updates the current TCI and cumulative TCI.
+func TestExtendTCI(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
Added as new testcase "Testcase-02"
It extends tci value to tagged TCI node.
Checks current and cumulative measurements are as expected.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,218 @@
-		t.Fatalf("GetTaggedTCI returned %v, want %v", err, StatusBadTag)
+// Check whether the ExtendTCI command with derived child context.
+func TestExtendTciOnDerivedContexts(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
Added a new testcase "Testcase-03", that creates a derived context, extrends TCI values to it and checks TCI_CURRENT and TCI_CUMULATIVE values

hpya93 commented on verification/verification.go:
```diff
@@ -59,6 +65,8 @@ var AllTestCases = []TestCase{
 	GetCertificateChainTestCase,
 	TagTCITestCase,
+	ExtendTCITestCase,
+	ExtendDerivedTciTestCase,
```
Reverted the testcases order in this list, added newly created testcases (that were split from TagTCI testcase)

hpya93 replied:
Reverted the ordering of testcases

hpya93 commented on verification/verification_test.go:
```diff
@@ -8,8 +8,6 @@ import (
 
-var testTargetType string
-
 // This will be called before running tests, and it assigns the socket path based on command line flag.
```
Unused variable

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,239 @@
-		t.Fatalf("GetTaggedTCI returned %v, want %v", err, StatusBadTag)
+// Check whether the ExtendTCI command with derived child context.
+func TestExtendTciOnDerivedContexts(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
Split into separate testcase as suggested in review comments.
Creates a child context, extends TCI and checks cumulative measurements.
Cleans up the child (derived) context and restores parent context as default context for subsequent tests.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,239 @@
+
+// Check whether the ExtendTCI command updates the current TCI and cumulative TCI.
+func TestExtendTCI(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
Split into new testcase as per review comment.
It extends TCI to existing context and verifies the current/cum measurements.

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,239 @@
+
+// Check tagTCI command with default context handle.
+func TestTagTCI(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
Has added general checks like 
- Tagging default context
- Check the handle returned by tagging Default context (must be unchanged)
- Error code of retagging a TaggedTCI node
- Error code of retrieving a non-existent tag etc

hpya93 commented on verification/go.mod:
```diff
@@ -11,7 +11,7 @@ require (
 	github.com/weppos/publicsuffix-go v0.30.1-0.20230422193905-8fecedd899db // indirect
 	golang.org/x/crypto v0.7.0 // indirect
-	golang.org/x/exp v0.0.0-20230817173708-d852ddb80c63 // indirect
+	golang.org/x/exp v0.0.0-20230817173708-d852ddb80c63
```
Based on a warning in IDE

hpya93 commented on verification/certifyKey.go:
```diff
@@ -544,11 +546,6 @@ func buildVerifyOptions(t *testing.T, certChain []*x509.Certificate) x509.Verify
 	return opts
 }
 
```
Unwanted function, not used anywhere

hpya93 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,239 @@
+		t.Errorf("Child node's cumulative TCI %x, expected %x", childExtendTci.CumulativeTCI, wantCumulativeTCI)
+	}
+
+	// Clean up derived context and restore default context handle for subsequnt tests
```
Code to cleanup derived context so that other tests iin the line are not affected and report unexpected errors.

jhand2 commented on verification/tagTCI.go:
```diff
@@ -3,55 +3,239 @@
-	if _, err := client.GetTaggedTCI(TCITag(98765)); !errors.Is(err, StatusBadTag) {
-		t.Fatalf("GetTaggedTCI returned %v, want %v", err, StatusBadTag)
+// Check whether the ExtendTCI command with derived child context.
+func TestExtendTciOnDerivedContexts(d TestDPEInstance, c DPEClient, t *testing.T) {
```
See https://github.com/chipsalliance/caliptra-dpe/issues/257. I would recommend moving these tests to be ExtendTci test and use CertifyKey to get the TCIs (instead of TagTCI).

hpya93 replied:
I have added a new function in certifyKey.go that parses MultiuTcbInfo, gets cumulative/current values from FWID array and cross checks with hashed values.
Also, the new tests in TagTCI.go are moved to ExtendTci.go.
Please let me know the comments!

hpya93 commented on verification/extendTCI.go:
```diff
@@ -0,0 +1,230 @@
+// Licensed under the Apache-2.0 license
+
+package verification
```
New file created to move new tests from TagTci.go file

hpya93 commented on verification/extendTCI.go:
```diff
@@ -0,0 +1,230 @@
+
+	// Cross-check current and cumulative measurement by GetTaggedTCI
+	verifyMeasurementsByGetTaggedTCI(c, t, defaultTci, tciValue, hasher)
+
```
Manually hashes the TCI values and cross checks with GetTaggedTci command.

hpya93 commented on verification/extendTCI.go:
```diff
@@ -0,0 +1,230 @@
+
+	// Cross-check current and cumulative measurement by CertifyKey
+	verifyMeasurementsByCertifyKey(c, t, handle, defaultTci, tciValue, hasher)
+}
```
Added a new test, instead of using GetTaggedTCI it uses CertifyKey to get TCI_CURRENT/TCI_CUMULATIVE. 
- Get leaf cert from CertifyKey
- Parse MultiTcbInfo 
- Get FWID array and use it to cross check

hpya93 commented on verification/extendTCI.go:
```diff
@@ -0,0 +1,230 @@
+
+// Check whether the ExtendTCI command with derived child context.
+func TestExtendTciOnDerivedContexts(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var err error
```
- Get the default context and cross-verify current/cum by recalculating the hash.
- Create/Derive a child context with supplied TCI value and cross-verify current/cum by recalculating the hash.
- Extend value to derived context  and cross-verify current/cum by recalculating the hash.

hpya93 commented:
Hi @jhand2 
The continuous integration job fails with warning in Rust code. 
I could run ci.sh in my machine without issues and all rust/go tests PASS.
Can you please guide?

jhand2 commented:
> Hi @jhand2 Have addressed all comments and updated the conversation(s). The continuous integration job fails with warning in Rust code. I could run ci.sh in my machine without issues and all rust/go tests PASS. Can you please guide?

The issue is that the CI is using a newer Rust toolchain which produces a lot of new clippy errors. This should fix the problem: https://github.com/chipsalliance/caliptra-dpe/pull/258

hpya93 commented:
> > Hi @jhand2 Have addressed all comments and updated the conversation(s). The continuous integration job fails with warning in Rust code. I could run ci.sh in my machine without issues and all rust/go tests PASS. Can you please guide?
> 
> The issue is that the CI is using a newer Rust toolchain which produces a lot of new clippy errors. This should fix the problem: #258

Hi @jhand2, 
JFYI it seems that check in PR 258 failed.
I locally fixed this warning in openssl.rsand tried in my machine.  
```
error: length comparison to zero
  --> crypto/src/openssl.rs:23:25
   |
23 |         let e_code = if s.len() > 0 {
   |                         ^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!s.is_empty()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero
   = note: `-D clippy::len-zero` implied by `-D warnings`
```

It gave warnings in rust build and ultimately did NOT build the simulator
```
# all "cargo build" commands in ci.sh is giving this warning
++ cargo clippy --manifest-path simulator/Cargo.toml -- --deny=warnings
warning: /home/haripriya/Hari/Caliptra/workspace/jhand2/caliptra-dpe/tools/Cargo.toml: unused manifest key: bin.0.target
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s


```
I have never seen this warning earlier when CICD was working.

hpya93 commented:
> > Hi @jhand2 Have addressed all comments and updated the conversation(s). The continuous integration job fails with warning in Rust code. I could run ci.sh in my machine without issues and all rust/go tests PASS. Can you please guide?
> 
> The issue is that the CI is using a newer Rust toolchain which produces a lot of new clippy errors. This should fix the problem: #258

Hi @jhand2, I have synced the branch to include your fix, CICD job completes now without issues, Can this be merged?
